doc/source/reference/internals.code-explanations.rst

```diff
@@ -188,139 +188,85 @@ written to so that structured array field setting works more naturally
-being placed into the array so that it will only work if the values
-used for setting have a shape that is "broadcastable" to the shape
-implied by the indexing object.
+All python indexing operations ``arr[index]`` are organized by first preparing
+the index and finding the index type. The supported index types are:
+* integer
+* newaxis
+* slice
+* ellipsis
+* integer arrays/array-likes (fancy)
+* boolean (single boolean array); if there is more than one boolean array as
+  index or the shape does not match exactly, the boolean array will be
+  converted to an integer array instead.
+* 0-d boolean (and also integer); 0-d boolean arrays are a special
+  case which has to be handled in the advanced indexing code. They signal
+  that a 0-d boolean array had to be interpreted as an integer array.
+
+As well as the scalar array special case signaling that an integer array
+was interpreted as an integer index, which is important because an integer
+array index forces a copy but is ignored if a scalar is returned (full integer
+index). The prepared index is guaranteed to be valid with the exception of
+out of bound values and broadcasting errors for advanced indexing. This
+includes that an ellipsis is added for incomplete indices for example when
+a two dimensional array is indexed with a single integer.
+
+The next step depends on the type of index which was found. If all
+dimensions are indexed with an integer a scalar is returned or set. A
+single boolean indexing array will call specialized boolean functions.
+Indices containing an ellipsis or slice but no advanced indexing will
+always create a view into the old array by calculating the new strides and
+memory offset.  This view can then either be returned or, for assignments,
+filled using :cfunc:`PyArray_CopyObject`. Note that `PyArray_CopyObject`
+may also be called on temporary arrays in other branches to support
+complicated assignments when the array is of object dtype.
+
+Advanced indexing
+-----------------
+
+By far the most complex case is advanced indexing, which may or may not be
+combined with typical view based indexing. Here integer indices are
+interpreted as view based. Before trying to understand this, you may want
+to make yourself familiar with its subtleties. The advanced indexing code
+has three different branches and one special case:
+* There is one indexing array and it, as well as the assignment array, can
+  be iterated trivially. For example they may be contiguous. Also the
+  indexing array must be of `intp` type and the value array in assignments
+  should be of the correct type. This is purely a fast path.
+* There are only integer array indices so that no subspace exists.
+* View based and advanced indexing is mixed. In this case there is a "subspace"
+  defined by the view based indexing and having the same functionality. For
+  example for ``arr[[1, 2, 3], :]`` the subspace is defined by ``arr[0, :]``.
```

Comment:
Oh, you worked on this. Thanks, though I hate to pass it on.

Well, it was supposed to say that you iterate the advanced indices and what is left is the "subspace" which moves around:
```
arr = np.ones((3, 4, 5))
arr[indxarr1, :, indxarr2]
```
Then this is much like:
```
res[0, :] = arr[[0], :, indx1[0]]
res[1, :] = arr[[1], :, indx1[1]]
res[2, :] = arr[[2], :, indx1[2]]
…
```
If you look at the strides, you get that each of that is:
```
arr[0, :, 0] + pointer_offset_according_to_arrays
```
so that `arr[0, :, 0]` is the "subspace" and the offset to where we are right now is defined by the current advamced index. Not sure if this all helps, but maybe the long example where you just unravel the advanced indexing helps a lot.